### DISCUSSION TITLE
Testing GraphRAG with other LLMs

### DISCUSSION THREAD

**Discussion by alexchaomander** [microsoft/graphrag#321] [Ideas]
The team primarily built GraphRAG with the GPT4-family of models and current prompts have been tested to work well with GPT4-o.

We'd love to see how GraphRAG works with open models like phi3, llama, mistral, etc.

Share your ideas, experiments, and experiences here!

**Comment by pradhandebasish2046** [comment]
Yes I am also looking forward to see the performance with other models

**Comment by namin** [comment]
I got the [Get Started](https://microsoft.github.io/graphrag/posts/get_started/) guide working entirely locally on Apple Silicon.

## Setup

### Toolio

I used https://github.com/OoriData/Toolio as my OpenAI-compatible server for text completions, because it uses Apple's MLX and supports JSON schemas.

I started the server with:

```shell
toolio_server --model=mlx-community/Hermes-2-Theta-Llama-3-8B-4bit
```

I also used a model with a larger context to satisfy the community report of graphrag:

```shell
toolio_server --model=mlx-community/Llama-3-8B-Instruct-1048k-4bit
```

I changed the `llm: api_base` setting to `http://127.0.0.1:8000/v1/`.

### open-text-embeddings

For embeddings, I used [open-text-embeddings](https://github.com/limcheekin/open-text-embeddings).

I started the server with:

```shell
PORT=8080 VERBOSE=1 MODEL=BAAI/bge-large-en python -m open.text.embeddings.server
```

I changed the `embeddings: llm: api_base` setting to `http://127.0.0.1:8080/v1/`.

With this, I was able to get the indexing. I think it took around 2 hours.

I was also able to run the global and local search queries, after resolving some tweaks which I describe next.

## Issues

### Enhancement: JSON Schema defined informally instead of formally

Toolio can work with JSON schemas and enforce them. I added the following schema:

```python
MAP_SYSTEM_JSON = """
{
  "type": "object",
  "properties": {
    "points": {
      "type": "array",
      "contains": {
        "type": "object",
        "properties": {
          "description": { "type": "string" },
          "score": { "type": "number" },
        },
        "required": ["description", "score"],
      },
      "minContains": 5,
    },
  },
  "required": ["points"],
}
"""
```

And then added a `"schema"` property set to that in the `"response_format" map_llm_params` in `./query/structured_search/global_search/search.py`.

I suspect whenever GraphRAG informally prompts for some JSON, the effectiveness with Toolio could be improved by passing the schema formally.

### Some bug with `streaming=True`

I had some issue seeing any response when LLM generation used `streaming=True`, so I changed those in `*_search/search.py` to `False`.

### `tiktoken` and non-OpenAI embeddings

Somehow, the open-text-embeddings server complains about the chunking using `tiktoken`, saying it expects a `string` not a `list` of `numbers`. Setting the `token_encoder` to `None`, and not using an encoder when `None` in `chunk_text` (in `./query/llm/text_utils.py`) seems to have resolved the issue for me.

```python
def chunk_text(
    text: str, max_tokens: int, token_encoder: tiktoken.Encoding | None = None
):
    """Chunk text by token length."""
    if token_encoder is None:
        tokens = text.split()
    else:
        tokens = token_encoder.encode(text)  # type: ignore
    chunk_iterator = batched(iter(tokens), max_tokens)
    yield from chunk_iterator
```

### Issue: special tokens within JSON generation

I did run into an issue with the large context llama that special tokens appeared amidst the JSON causing decoding errors. I did a brutal cleanup as follows:

```python
                search_response = search_response.replace("<|start_header_id|>assistant<|end_header_id|>", "")
                search_response = search_response.split('<|eot_id|>')[0]
```

But that was also a shame, because I discarded all the alternate responses.

## What's next?

I haven't evaluated how well the system work -- my first victory was to get it running!

I might add a JSON schema for Toolio for each prompt that requires JSON generation.

I am not sure whether there's anything to push upstream at the moment: all the issues I hit could probably be solved by improving the OpenAI-compatible servers.

**Reply by streamside7** [reply]
What an awesome post! Thanks!

**Comment by xxWeiDG** [comment]
It's work on chatglm4-9b-chat with   xinference

**Comment by shaoqing404** [comment]
I tried deepseek as a lower-level alternative to gpt4o, and mentioned in the issue that many utf8 and json problems (before 0.2.1) will not be encountered in this model

**Comment by timothymeyers** [comment]
I've been experimenting with @TheAISingularity's [repo here](https://github.com/TheAiSingularity/graphrag-local-ollama).

This is not an endorsement, but it seems to work well leveraging ollama locally on Apple Silicon to try out different models.

**Comment by l4b4r4b4b4** [comment]
I am using a custom vLLM server fork incl. openai compatible tool & function calling as well as an embeddings endpoint  using funtionary-3.1-small (llama3.1-8B) AWQ quant on my RTX 3080 and indexing as well as global search work amazingly well. Both with `JSON=True` and `JSON=True`. 

Will try with bigger versions of the functionary model family in the near future...

Local search does not work. But did not look to deeply, why that is the case.

**Reply by shaoqing404** [reply]
本地搜索的问题你可以尝试现在global中进行搜索，积累社区以后再回到local进行检索。你会有不一样的体验。

**Comment by nileshtrivedi** [comment]
I think it would be great if GraphRAG project was built on top of another Microsoft project, [Autogen](https://github.com/microsoft/autogen). This way, interoperability with various LLM providers would need to be implemented only in one place.

Another approach is to use existing open-source LLM interoperability libraries like [LiteLLM](https://github.com/BerriAI/litellm).

**Comment by ntsarb** [comment]
Please check my relevant post here:
https://github.com/microsoft/graphrag/discussions/1960

I have publicly shared the code because I am working on this entirely on my personal time, typically after I put my  son to bed in the evening, and any help to review and test the changes may help to confirm whether the approach may work (or not).  If it does, it would be great if this is incorporated into the original, main branch.

**Comment by kinthaiofficial** [comment]
GraphRAG with non-OpenAI LLMs is worth the effort — the entity extraction and community summarization quality does vary across models, but the graph structure itself provides quality improvement even with weaker models.

A few findings from testing with alternative LLMs:

**Entity extraction quality is the critical bottleneck** — the initial indexing phase (entity extraction, relationship detection) is where model quality matters most. A weaker model here produces a sparse graph that loses edges. We found 70B models (Llama 3, Qwen 2.5) produce acceptable graphs; smaller models degrade noticeably.

**Community summarization tolerates smaller models** — once the graph exists, summarizing communities is a simpler task that 13B-34B models handle well. The compression task is more forgiving than the extraction task.

**Prompt adaptation is necessary** — GraphRAG's default prompts are tuned for GPT-4 instruction following. For other models, especially those with different system prompt handling, adapting the extraction prompts is usually needed. The key prompts are the entity extraction and relation extraction templates.

**Embedding model matters independently** — the embedding model for vector search within GraphRAG is separable from the LLM. Using a strong embedding model (text-embedding-3-large or equivalent) compensates partially for weaker generation.

**Ollama/LocalAI as the endpoint** — for local models, wrapping with an OpenAI-compatible endpoint (Ollama's API) is the easiest path. The `llm_type: openai_chat` config with a local endpoint works once you set the right base URL.

We've been using GraphRAG-style knowledge graph extraction for KinthAI's agent memory layer: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

Which model are you testing against — is it a hosted alternative or a local model?
